### ISSUE TITLE
Implement batch deletion of S3 sstables

### ISSUE THREAD

**Issue by xemul** [scylladb/scylladb#13567]
There's the need to atomically delete a bunch of sstables. For disk-backed sstables this is implemented inside sstable_directory::delete_atomically() which puts sstables names into deletion logs, then drops them one by one. If this crashes int the middle, subsequent restart would pick the deletion log and drop the remainder.

Fos s3-backed sstables deletion log is not nice. At the same time updating ownership table for a bunch of sstables would require batch statement, but (quoting Raphael)

> My concern about batches is that compactions like reshape have hundreds of entries deleted together. If you have 96 shards, you'll have 96 batches, with hundreds of operations each, happening roughly at the same time. The alternative is to implement deletion logs in S3, perhaps storing an entry for the log in sstables_registry and reusing existing fields like format and version to identify it.

**Comment by xemul** [comment]
How about this, @raphaelsc ?

Add new column to sstables registry called, say, `batch_deletion` containing string. The value indicates the batch deletion cookie.

The delete_atomically() then executed this _sequence_:
1. generates new ID
2. walks the sstables to get dropped and updates each with the same cookie
3. walks the entries and sets status to "deleting"
4. calls regular .wipe() on entries (wipe removes entries from table at the end)

On boot all entries with non-empty `batch_deletion` markers are collected and grouped by cookie value.

If _all_ entries in the batch are in "deleting" state it means, that the batch was at step 4, so it should go ahead and complete it.

Else, the batch was in steps 1, 2 or 3 and we don't even know in which, so the status should be restored back to "sealed" and batch deletion cookie should be dropped. This is equivalent to the case when on-disk deletion log was created, but wasn't flushed or renamed to its permanent location.

In any case after reboot the whole batch run will either be removed altogether, or will be rolled back to its original state.

Additionally, the cookie can become `tuple<string, int>` where the number denotes how many entries were in the batch on start. This would allow to tell steps 1 and 2 from 3 -- if boot collects a batch run with some entries in "sealed" state, but their number matches the cookie's count, this means we were in step 3 and can proceed with wiping as well.

**Comment by avikivity** [comment]
> There's the need to atomically delete a bunch of sstables. For disk-backed sstables this is implemented inside sstable_directory::delete_atomically() which puts sstables names into deletion logs, then drops them one by one. If this crashes int the middle, subsequent restart would pick the deletion log and drop the remainder.
> 
> Fos s3-backed sstables deletion log is not nice. At the same time updating ownership table for a bunch of sstables would require batch statement, but (quoting Raphael)
> 
> > My concern about batches is that compactions like reshape have hundreds of entries deleted together. If you have 96 shards, you'll have 96 batches, with hundreds of operations each, happening roughly at the same time. The alternative is to implement deletion logs in S3, perhaps storing an entry for the log in sstables_registry and reusing existing fields like format and version to identify it.

I don't see a problem with those batches. Batches into raft-managed system tables are simple since they don't cross node or shard boundaries like vnode batches (and in any case these are single-partition batches). They should work well.

**Comment by raphaelsc** [comment]
> > There's the need to atomically delete a bunch of sstables. For disk-backed sstables this is implemented inside sstable_directory::delete_atomically() which puts sstables names into deletion logs, then drops them one by one. If this crashes int the middle, subsequent restart would pick the deletion log and drop the remainder.
> > Fos s3-backed sstables deletion log is not nice. At the same time updating ownership table for a bunch of sstables would require batch statement, but (quoting Raphael)
> > > My concern about batches is that compactions like reshape have hundreds of entries deleted together. If you have 96 shards, you'll have 96 batches, with hundreds of operations each, happening roughly at the same time. The alternative is to implement deletion logs in S3, perhaps storing an entry for the log in sstables_registry and reusing existing fields like format and version to identify it.
> 
> I don't see a problem with those batches. Batches into raft-managed system tables are simple since they don't cross node or shard boundaries like vnode batches (and in any case these are single-partition batches). They should work well.

You're right. Single-partition batches allow for unlogged ones.

**Comment by raphaelsc** [comment]
> How about this, @raphaelsc ?
> 
> Add new column to sstables registry called, say, `batch_deletion` containing string. The value indicates the batch deletion cookie.
> 
> The delete_atomically() then executed this _sequence_:
> 
> 1. generates new ID
> 2. walks the sstables to get dropped and updates each with the same cookie
> 3. walks the entries and sets status to "deleting"
> 4. calls regular .wipe() on entries (wipe removes entries from table at the end)
> 
> On boot all entries with non-empty `batch_deletion` markers are collected and grouped by cookie value.
> 
> If _all_ entries in the batch are in "deleting" state it means, that the batch was at step 4, so it should go ahead and complete it.
> 
> Else, the batch was in steps 1, 2 or 3 and we don't even know in which, so the status should be restored back to "sealed" and batch deletion cookie should be dropped. This is equivalent to the case when on-disk deletion log was created, but wasn't flushed or renamed to its permanent location.
> 
> In any case after reboot the whole batch run will either be removed altogether, or will be rolled back to its original state.
> 
> Additionally, the cookie can become `tuple<string, int>` where the number denotes how many entries were in the batch on start. This would allow to tell steps 1 and 2 from 3 -- if boot collects a batch run with some entries in "sealed" state, but their number matches the cookie's count, this means we were in step 3 and can proceed with wiping as well.

That would certainly work. Interesting technique btw :-) 

But as Avi pointed out, batches will not be a problem. SSTables will share the same partition key in the system table. That will simplify things a lot more as we just need to emit one status update for sstables being deleted together. On restart, we proceed with wipe of all entries that have the removing status.

One question, if we have multiple S3 prefixes for a given table, the logical prefix (which is represented as primary key in the registry table) will still be the same for all SSTables, right? I see that you did some effort for decoupling physical location from logical one. That's great.

**Comment by xemul** [comment]
> SSTables will share the same partition key in the system table. 

Will they? The partition key is sstable path ( `${keyspace_name} / ${table_name} _ ${table_id}` ). If all sstables in delete_atomically() belong to the same table, then yes -- they share the partition, but is that guaranteed that delete_atomically() is only called for a vector of sstables belonging to one table?

> One question, if we have multiple S3 prefixes for a given table, the logical prefix (which is represented as primary key in the registry table) will still be the same for all SSTables, right? 

Yes. The partition key in the table is "ks/cf-id" thing described above that's obtained from the table. Generation value is clustering key. And the S3 prefix is just a cell.

> I see that you did some effort for decoupling physical location from logical one. That's great.

:)

**Comment by raphaelsc** [comment]
> > SSTables will share the same partition key in the system table.
> 
> Will they? The partition key is sstable path ( `${keyspace_name} / ${table_name} _ ${table_id}` ). If all sstables in delete_atomically() belong to the same table, then yes -- they share the partition, but is that guaranteed that delete_atomically() is only called for a vector of sstables belonging to one table?
> 

Yes, we can always assume that. It's currently enforced in sstable_directory::delete_atomically(). In the upcoming sstables_manager::delete_atomically() (introduced in https://github.com/scylladb/scylladb/pull/13562) we can still enforce it by checking that all sstables being deleted together share the same table's id.

**Comment by xemul** [comment]
Another idea (not that I'm trying to avoid batchlog-based implementation).

The need to atomically delete sstables 1, 2, ... N comes from compaction that squashed them into one sstable M. So atomic deletion makes sure we only have one of the following sets on disk:
1. { 1, 2, ... N }
2. { 1, 2, ... , N, M }
3. { M }

No other combinations are allowed. Said that, we can mark sstables 1 through N with a "dependant" field that's empty by default, but when compaction starts its job it marks the initial sstables with dependant = M, then writes sstable M, then seals it. Then deletes sstables 1, ... N. On boot meeting sstable(s) with dependant field set we'd need to check if the dependant target exists and is sealed and just remove the original sstable(s). If the dependant target doesn't exist or isn't sealed, the mark should be just dropped, the target would be removed on its own, because it's not sealed.

Thoughts?

**Comment by avikivity** [comment]
We don't care what's on disk, since the source-of-truth is the sstable table.

We don't scan the bucket, we scan the sstable table. So if some sstable object is left behind it will be ignored.

We'll need to be careful with garbage collection:
 - if a node restarts, it will have to delete any sstables that it owns and are marked "deleted" (and then purge the entry)
 - if a node goes away, the leader will have to delete all of its sstables (and then purge their entries)

**Comment by xemul** [comment]
> We don't care what's on disk, since the source-of-truth is the sstable table.
> 
> We don't scan the bucket, we scan the sstable table. So if some sstable object is left behind it will be ignored.

Agree, and the above suggestion is about manipulating the ownership table and only it. I guess I wasn't clear enough.

So, consider there are entries in the ownership table for sstables 1, 2, ... N. Next compaction comes and wants to a) compact those sstables into another one M and b) remove sstables 1 through N atomically. Atomically means that either all 1...N are removed or _not_ removed, not _part_ of it.

To get there the algo is like this. When opening sstable M, an entry about it is added to the ownership table. Also entries for 1 through N are updated with an extra column, say, "depends-on-uuid", the value is set to M. Then M is written to and sealed. Then sstables 1...N are deleted normally.

After reboot when scanning the ownership table all entries with non-empty "depends-on-uuid" column are grouped together and the entry with uuid == M is checked. If it's in "sealed" state, then the whole group is removed. If it's in any other state, then M is removed, but entries 1...N have its "depends-on-uuid" column updated with empty value.

**Comment by avikivity** [comment]
We don't need to be so strict. It's okay to have 1..N,  M in the sstable table. When compaction completes, it can add M, then mark 1..N as being deleted, delete them, then remove the entries.

**Comment by avikivity** [comment]
> Another idea (not that I'm trying to avoid batchlog-based implementation).
> 
> The need to atomically delete sstables 1, 2, ... N comes from compaction that squashed them into one sstable M. So atomic deletion makes sure we only have one of the following sets on disk:
> 
> 1. { 1, 2, ... N }
> 2. { 1, 2, ... , N, M }
> 3. { M }
> 
> No other combinations are allowed. Said that, we can mark sstables 1 through N with a "dependant" field that's empty by default, but when compaction starts its job it marks the initial sstables with dependant = M, then writes sstable M, then seals it. Then deletes sstables 1, ... N. On boot meeting sstable(s) with dependant field set we'd need to check if the dependant target exists and is sealed and just remove the original sstable(s). If the dependant target doesn't exist or isn't sealed, the mark should be just dropped, the target would be removed on its own, because it's not sealed.
> 
> Thoughts?

I don't understand where batchlog comes from. We can delete all the entries in a single Raft transaction.
